Type work for some handlers and stuff related to the compiler-finder

jeremy-rifkin opened the pull request:
Types are needing a lot of work for a handful of files in lib/ and test/

partouf commented:
Seems to me that changing the type for `compilerInfo.exe` to non-optional makes more sense than to add unwrap to all the things. Just need to make sure it is always set during find/discovery/test.

RubenRBS commented on lib/common-utils.ts:
```diff
@@ -36,3 +36,11 @@ export function keys<K extends string | number | symbol>(o: Partial<Record<K, an
+// arr.filter(x => x !== null) returns a (T | null)[] even though it is a T[]
+// Apparently the idiomatic solution is arr.filter((x): x is T => x !== null), but this is shorter (and the type
+// predicate also isn't type checked so it doesn't seem safe to me)
+export function remove<U, V extends U>(arr: U[], v: V) {
```
🧙

jeremy-rifkin commented on lib/compiler-finder.ts:
```diff
@@ -129,14 +137,14 @@ export class CompilerFinder {
                                             if (typeof compiler.alias == 'string') compiler.alias = [compiler.alias];
                                             // End fixup
-                                            compiler.exe = null;
+                                            compiler.exe = '/dev/null';
```
Per chatting with @partouf. I think it's slightly better to use '/dev/null' here instead of an empty string or leaving it undefined/null, it should give us an easy to track down error if the compiler ever tries to exec it.

jeremy-rifkin commented on test/compiler-finder-tests.js:
```diff
@@ -119,7 +119,7 @@ describe('Compiler-finder', function () {
     it('should not hang for undefined groups (Bug #860)', () => {
         const finder = new CompilerFinder({}, compilerProps, properties.fakeProps({}), {}, optionsHandler);
-        return finder.getCompilers().should.eventually.have.lengthOf(2);
+        return finder.getCompilers().should.eventually.have.lengthOf(1);
```
Before the changes I've made to CompilerFinder `finder.getCompilers()` here was returning a `[{ ... actual compiler ... }, []]`. Now it just returns `[{ ... actual compiler ... }]`.

jeremy-rifkin commented on lib/base-compiler.ts:
```diff
@@ -305,7 +308,7 @@ export class BaseCompiler implements ICompiler {
 
     getRemote() {
-        if (this.compiler.exe === null && this.compiler.remote) return this.compiler.remote;
+        if (this.compiler.exe === '/dev/null' && this.compiler.remote) return this.compiler.remote;
```
If this PR introduces bugs, I wouldn't be surprised if it was stuff like this. As far as I can tell this is the only check of CompilerInfo.exe.

partouf replied:
I think it might be best here to make `this.compiler.remote` the only condition

jeremy-rifkin commented:
Ok - ready for review
There are some behavior changes in the compiler finder, mainly related to how nested arrays are flattened and how null entries are removed. For remote compilers, `compiler.exe` is set to `'/dev/null';` instead of null. Also removed instances of _.js.
I think most of the other changes are pretty straightforward. I made some fields optional in the CompilerInfo type and added a number of unwraps to deal with those. I'll want to test thoroughly in staging before this is deployed.

jeremy-rifkin commented:
Tested briefly on staging, things look good to me
